### ISSUE TITLE
Research plan: prerequisite-based new-skill acquisition

### ISSUE THREAD

**Issue by kobros-tech** [kobros-tech/skill-cloning#3]
## Goal

Build and evaluate a small, rigorous skill-acquisition system whose primary purpose is **reliably acquiring new skills while retaining previously acquired skills**.

> **The AI's ability to acquire a new skill is the primary outcome. Prior skills and cloning are valuable only insofar as they make acquisition more reliable or more efficient, and the system should retain previously acquired skills while learning new ones.**

Cloning is a means, not the goal. The system must always retain a genuine **scratch-learning fallback**. We do not need to force cloning to win.

---

# Current research question

The main hypothesis is:

> **A new skill should be acquired more quickly and/or more reliably when the AI has already acquired enough relevant prerequisite skills. The system should also retain previously acquired skills while acquiring new ones, rather than suffering substantial catastrophic forgetting.**

This is a hypothesis, not an assumption. The experiments must be capable of showing no benefit, negative transfer, or measurable forgetting.

The key experimental idea is:

> **Keep the target task fixed and vary the prior-skill history.**

This is more informative than simply collecting unrelated source→target pairs.

---

# Acquisition routes

For every new target, the system should support:

1. **Reuse** — an existing skill already solves the target; zero adaptation training.
2. **Clone + adapt** — an existing skill is useful initialization, then is trained on the target.
3. **Scratch** — fresh initialization and target training.

The corrected PR #2 controller must be used wherever the experiment makes an actual controller decision. Do **not** use a hand-written `always clone` path as the main acquisition experiment.

For `reuse`, continue to require both:

- compatibility score ≥ `τ_solve`; and
- independent target solve accuracy ≥ 85%.

---

# PR history and current state

```text
main
  ├── PR #1: stopping-rule confound                 [merged]
  ├── PR #2: squares + calibration/controller fix   [merged]
  └── PR #4: relatedness-pairs experiment           [current]
         ↓
      revise PR #4 to implement the experiment below
         ↓
      PR #5: retention / catastrophic-forgetting experiment
         ↓
      later statistical analysis / final paper
```

**Important:** PR #4 contains the earlier four-pair relatedness experiment. Those results are useful transitional evidence and should not be thrown away, but the fixed-target prerequisite experiment is the main implementation for this issue.

PR #4 currently compares:

| Pair | Mean speedup |
|---|---:|
| Multiplication → Squares | 1.26× |
| Multiplication → Powers | 2.31× |
| Addition → Subtraction | 0.33× |
| Addition → Multiplication | 1.12× |

These results are useful for motivating the fixed-target experiment, especially the observation that transfer benefit is not necessarily monotonic in the frozen compatibility score.

---

# Next implementation: prerequisite matrix

For each target, compare the **same target** under different prior-skill histories.

### Initial matrix

| Target | Prior skills | Acquisition |
|---|---|---|
| Subtraction | None | ? |
| Subtraction | Addition | ? |
| Subtraction | Addition + Multiplication | ? |
| Division | None | ? |
| Division | Addition | ? |
| Division | Addition + Multiplication | ? |
| Squares | None | ? |
| Squares | Addition | ? |
| Squares | Addition + Multiplication | ? |
| Powers | None | ? |
| Powers | Addition | ? |
| Powers | Addition + Multiplication | ? |

This matrix is the minimum starting design. Any omitted cell must be explained.

The most important comparisons are:

```text
Subtraction:
    scratch
    addition → subtraction
    addition → multiplication → subtraction

Division:
    scratch
    addition → division
    addition → multiplication → division
```

Existing multiplication→squares and multiplication→powers results may be reused as historical evidence, but new runs should be performed where necessary so that the histories are directly comparable.

---

# Scientific hypotheses

### H1 — Relevant prerequisite accumulation helps

A target should be acquired faster and/or with higher fixed-budget success after enough relevant prerequisite skills than from scratch.

### H2 — Partial prerequisites may provide less benefit

A partial prerequisite history may help less than a more complete relevant history.

### H3 — Missing prerequisites may remove the benefit

If an important prerequisite has not been acquired, cloning an earlier skill may provide little advantage over scratch.

### H4 — Negative transfer is possible

An incomplete or poorly matched prior skill may make acquisition slower than scratch. This must be measured, not manufactured.

### H5 — Relevance matters, not simply the number of prior skills

Where computationally reasonable, include an irrelevant-prior-skill control so that we can distinguish:

> more prior training always helps

from:

> relevant prior knowledge helps.

### H6 — Previously acquired skills should be retained

After learning one or more new skills, performance on previously acquired skills should remain stable within the predefined retention criterion. The experiment must measure this directly rather than infer retention from acquisition success.

Do not claim that a curriculum order proves a formal mathematical prerequisite. The experiment only tests whether prior learned representations are useful for acquisition.

---

# Experimental protocol

Every condition must use the same:

- model architecture;
- optimizer;
- learning rate and other training hyperparameters;
- target training procedure;
- convergence criterion;
- fixed training budget;
- evaluation procedure;
- seed set, with matched seeds where appropriate.

Keep these data roles separate:

1. target training data;
2. compatibility probe data;
3. solve/accuracy data used by the reuse gate;
4. held-out final evaluation data.

For the retention experiment, additionally maintain a **retention-evaluation set for each previously acquired skill**, kept separate from all training data and never used for optimization or controller decisions.

Never use held-out or retention-evaluation data to select thresholds, budgets, stopping rules, or model-selection decisions.

The experiment should use the same corrected controller from PR #2 whenever controller decisions are part of the measured system.

---

# Primary outcome: acquisition reliability

The first question is:

> **Can the AI acquire the target skill within a predeclared fixed training budget?**

For each target/history/seed record whether acquisition succeeds.

Report success rates such as:

| Target | History | Success |
|---|---|---:|
| Division | Scratch | ?/15 |
| Division | Addition | ?/15 |
| Division | Addition + Multiplication | ?/15 |

A difference in success rate is more important than a small average speedup.

If all conditions reach 100%, that is still useful: then compare training cost/efficiency.

---

# Secondary outcome: acquisition efficiency

Measure:

- convergence steps/epochs;
- training cost;
- early learning curves;
- fixed-budget performance;
- final held-out performance.

Do not treat a small speedup such as 1.26× as automatically practically important.

A particularly strong result would be:

```text
Scratch:                    fails within budget
Partial prerequisites:      sometimes succeeds
Complete prerequisites:     reliably succeeds
```

Another useful result would be a large reduction in training cost while maintaining equal success.

---

# Retention / catastrophic-forgetting experiment — PR #5

The purpose of PR #5 is to test the second part of the research goal directly: **can the system acquire new skills without substantially losing previously acquired skills?**

This should be a focused extension of the existing task family rather than a new unrelated benchmark. Reuse the existing mathematical skills/tasks where possible (e.g. addition, subtraction, multiplication, division, squares, and powers) so that the retention result is directly connected to the acquisition experiment.

### Basic protocol

For each seed/sequence:

1. Train/acquire an initial skill and measure its pre-learning evaluation accuracy.
2. Acquire one or more subsequent skills using the same acquisition framework and controller where applicable.
3. After each new skill is acquired, re-evaluate every previously acquired skill on its separate retention-evaluation set.
4. Record both the new-skill acquisition result and the old-skill retention result.
5. Repeat across matched seeds so that pre/post comparisons are paired.

Example sequence:

```text
addition → measure addition
addition → multiplication → re-test addition and multiplication
addition → multiplication → division → re-test all three
```

Additional sequences can use subtraction, squares, and powers where computationally reasonable.

### Retention metrics

For each previously acquired skill `s`, record:

- pre-new-learning accuracy: `A_pre(s)`;
- post-new-learning accuracy: `A_post(s)`;
- absolute retention change: `ΔA(s) = A_post(s) - A_pre(s)`;
- relative retention ratio where meaningful: `A_post(s) / A_pre(s)`;
- whether the predefined retention criterion is met.

Do **not** choose a retention threshold after seeing the final results. The criterion must be declared before the full experiment.

Rather than inventing an arbitrary 2% or 5% cutoff, define the primary retention claim around the paired change and its uncertainty. A useful analysis should report the mean paired change, confidence interval (or bootstrap interval where appropriate), effect size, and the fraction of runs meeting the predefined practical-retention criterion.

A practical retention tolerance may be selected before running the full experiment based on measurement noise and the task's resolution; document the rationale rather than choosing it to make the result positive.

### What counts as evidence

A strong result would be:

```text
new skill acquired successfully
AND
previous skills remain within the predefined retention tolerance
AND
there is no statistically/materially meaningful systematic degradation across seeds.
```

If old-skill performance decreases materially, report **catastrophic forgetting or retention failure** honestly. Do not tune the experiment to avoid observing it.

This experiment does **not** prove universal absence of catastrophic forgetting. Any claim must be limited to the tested architecture, tasks, training procedure, budgets, and sequences.

---

# Required per-seed data

The experiment must produce enough information to reconstruct each acquisition path:

- target task;
- prior-skill history;
- seed;
- selected strategy (`reuse`, `clone`, `scratch`);
- source/parent skill where applicable;
- compatibility score;
- independent solve accuracy where applicable;
- adaptation training steps/epochs;
- fixed budget;
- acquisition success/failure;
- held-out target performance;
- fixed-budget performance.

For PR #5 additionally record:

- skill sequence;
- skill being re-evaluated;
- pre-new-learning accuracy;
- post-new-learning accuracy;
- absolute retention change;
- relative retention ratio where applicable;
- predefined retention criterion result.

CI should upload the complete per-seed CSV and a human-readable summary for both acquisition and retention experiments.

---

# Interpretation

Do not force a positive cloning or retention result.

All of the following are valid outcomes:

- scratch succeeds and cloning adds little value;
- cloning substantially reduces training cost;
- cloning increases fixed-budget success;
- partial prior knowledge gives little benefit;
- complete prerequisite knowledge gives a large benefit;
- incomplete prior knowledge produces negative transfer;
- an assumed prerequisite turns out not to matter;
- new skills are acquired while old skills remain stable;
- some sequences cause measurable forgetting;
- forgetting is task- or sequence-dependent.

The conclusion must follow the measurements.

---

# Implementation instructions for PR #4

The current coding step should revise PR #4 rather than immediately creating another unrelated acquisition experiment.

### Before coding

1. Inspect the current `experiments/relatedness_pairs.py` and existing experiment infrastructure.
2. Identify which functions/results can be reused.
3. Explain briefly how the current four-pair experiment is transformed into the fixed-target prerequisite experiment.

### Then implement

1. Add a reusable experiment runner for one `target + prior-history + seed` condition.
2. Implement the prerequisite matrix above.
3. Keep scratch as a genuine fresh-initialization baseline.
4. Use the corrected PR #2 controller for actual strategy decisions.
5. Add/update regression tests for the new experiment structure.
6. Update CI to run the new experiment and upload results.
7. Update `print_summary.py` and README/report documentation.
8. Run a small smoke test first.
9. Run the full seed set only after the smoke test passes.
10. Verify the generated results independently before considering PR #4 complete.

### Do not

- force cloning;
- use `always clone` as the main acquisition path;
- select thresholds/budgets from held-out or retention-evaluation data;
- silently replace old results without documenting the change;
- interpret the compatibility score as proof of transfer usefulness;
- claim that a task is a prerequisite merely because it appears earlier in the curriculum;
- claim absence of catastrophic forgetting without direct pre/post retention measurements.

---

# Implementation instructions for PR #5 — retention / catastrophic forgetting

PR #5 should be created **after PR #4's acquisition experiment is stable**, and should extend the same experimental infrastructure.

### Before coding

1. Inspect the finalized PR #4 acquisition runner and result schema.
2. Reuse the existing task definitions, model, training configuration, seeds, and evaluation conventions wherever possible.
3. Define the retention criterion and statistical analysis plan **before** running the full experiment.
4. Choose a small set of representative skill sequences first; expand only if the smoke test indicates the protocol is sound.

### Then implement

1. Add a reusable sequential-acquisition/retention runner.
2. Record pre-learning accuracy for every skill that will later be tested for retention.
3. Acquire the next skill using the established acquisition framework.
4. Re-test all previously acquired skills on independent retention-evaluation data after each acquisition stage.
5. Add regression tests for pre/post retention bookkeeping and data-role separation.
6. Update CI to run the retention experiment and upload complete per-seed results.
7. Update `print_summary.py` with retention summaries.
8. Add a concise report/README section describing the retention protocol and its limitations.
9. Run a smoke test before the full seed set.
10. Verify the generated pre/post results independently before considering PR #5 complete.

### PR #5 acceptance criteria

- Existing skills are explicitly re-evaluated after new-skill acquisition.
- Retention evaluation data are independent of training data.
- The retention criterion is declared before the full run.
- Results are paired by seed where appropriate.
- Both acquisition success and retention are reported.
- No claim of universal no-forgetting is made beyond the tested setting.

---

# Later statistical analysis / final paper

Once PR #4 and PR #5 are stable, a later PR should produce the final statistical analysis and paper-ready tables/figures:

### Acquisition

- acquisition success rates and uncertainty;
- convergence/training-cost distributions;
- paired clone-vs-scratch comparisons where applicable;
- effect sizes;
- confidence intervals / bootstrap intervals where appropriate;
- fixed-budget comparisons;
- learning curves;
- reliability vs. efficiency separately.

### Retention

- pre/post paired accuracy distributions;
- mean paired retention change;
- confidence/bootstrap intervals;
- effect sizes;
- proportion of runs meeting the predefined retention criterion;
- retention by skill and by acquisition sequence;
- learning/acquisition success and retention reported together.

Increase the number of seeds only when needed for informative uncertainty estimates.

---

# Final research story

The final report should be organized around **continual new-skill acquisition with retention**:

1. Can the AI acquire a new skill at all?
2. Can it correctly reuse a skill that already solves the target?
3. When reuse is impossible, does cloning provide a useful initialization?
4. Does the usefulness of cloning depend on having acquired the right prerequisites?
5. Can incomplete or mismatched prior knowledge cause negative transfer?
6. When does cloning materially improve reliability or efficiency compared with scratch?
7. **Can the AI acquire new skills while retaining previously acquired skills?**
8. **Under which tested sequences, if any, does catastrophic forgetting occur?**

A defensible final claim, if supported by the data, would be:

> **In a controlled small-scale continual-learning setting, previously acquired relevant skills can improve the reliability and/or efficiency of acquiring new skills, while a separate retention experiment tests whether previously acquired skills remain stable as new skills are learned. The system therefore supports reuse, clone-and-adapt, and scratch learning without assuming cloning is universally beneficial, and any retention claim is limited to the tested tasks, architecture, training procedure, and sequences.**

Do not claim a universal prerequisite hierarchy, universal cloning advantage, or universal absence of catastrophic forgetting from this small task set.

---

# Definition of completion

This issue is complete when we have:

- a reproducible new-skill acquisition pipeline;
- genuine scratch fallback;
- corrected reuse decisions;
- the fixed-target prerequisite matrix tested;
- acquisition success as the primary metric;
- efficiency as a secondary metric;
- possible negative transfer measured honestly;
- relevant-vs-irrelevant prior knowledge distinguished where practical;
- a direct sequential retention / catastrophic-forgetting experiment;
- predefined retention criterion and statistically interpretable pre/post comparisons;
- complete CI artifacts for acquisition and retention;
- statistically interpretable final results;
- explicit limitations and alternative explanations;
- paper-ready tables, figures, and documentation.
